### PR TITLE
Removed the Form interface

### PR DESCRIPTION
**Pull request by SOF3**
## Introduction

The Form interface is redundant. What PocketMine needs is just a JSON blob and a callback.

### Relevant issues

Previously discussed in #2865

### API changes
- Removed pocketmine\form\Form
- `Player->sendForm(array $data, Closure $callback)`
- `NetworkSession->onFormSubmit(int $id, array $data)`

## Backwards compatibility

API BC-breaking change.

## Follow-up

The `pocketmine\form\` namespace now only contains one class. This is a bit fishy.

### CODE REVIEW

File: src/pocketmine/Player.php [hunk 1]
```diff
@@ -2277,14 +2277,15 @@ public function sendTip(string $message){
...
 	 * Sends a Form to the player, or queue to send it if a form is already open.
 	 *
-	 * @param Form $form
+	 * @param array $data
+	 * @param \Closure $callback
 	 *
 	 * @throws \InvalidArgumentException
 	 */
-	public function sendForm(Form $form) : void{
+	public function sendForm(array $data, \Closure $callback) : void{
```
Comment (dktapps) [inline src/pocketmine/Player.php]: pls add a signature check here with `Utils::validateCallableSignature()`

File: src/pocketmine/Player.php [hunk 2]
```diff
@@ -2295,18 +2296,19 @@ public function sendForm(Form $form) : void{
...
 			return false;
 		}
 
 		try{
-			$this->forms[$formId]->handleResponse($this, $responseData);
+			$closure = $this->formHandlers[$formId];
+			$closure($this, $responseData);
 		}catch(FormValidationException $e){
-			$this->logger->critical("Failed to validate form " . get_class($this->forms[$formId]) . ": " . $e->getMessage());
+			$this->logger->critical("Failed to validate form: " . $e->getMessage());
```
Comment (dktapps) [inline src/pocketmine/Player.php]: might be better to use `Utils::getNiceClosureName()` here for debug info instead of removing it completely.

**Review by dktapps — COMMENTED** [review]
Concept ACK, I've made note of some minor nits.

**Comment by HimbeersaftLP** [comment]
Imo this reduces code readability because the Form interface made it more clear what you're passing to a function.

**Comment by SOF3** [comment]
Using your logic, the `Array.reduce(initialValue, reduceFn)` function should be changed into `Array.reduce(Reducer)`, e.g. they should have `[1, 2, 3].reduce(new MultiplyReducer())` rather than `[1, 2, 3].reduce(1, (a,b)=>a*b)`.
This apparently doesn't make sense. The sendForm function is a low level function; it should be called from form libraries, but directly on the user land. There is no reason to coerce two barely related parameters into a single object.

**Comment by dktapps** [comment]
To clarify the motivation for this: The form interface has 2 jobs:
1) being json-encodable
2) providing a result handling function

These two jobs are unrelated in functionality. Having them bound together creates a needless complexity.

For example, I wanted to implement a form type which was translated on the fly a la #2865 . This would require that the `Form` interface be changed to allow a `Player` context to be passed.
As it turns out, it's possible to do this with the following glue code:
https://media.discordapp.net/attachments/373199722573201410/589111049962258455/unknown.png
However, this glue code is only necessary because of a needless restriction in the API that enforces the JSON data provider and the response handler must both be in the same class.

The bottom line is that the Form interface forces you to do two jobs at once in the same, which is an unnecessary limitation.

**Comment by HimbeersaftLP** [comment]
oh

**Comment by dktapps** [comment]
The main thing that stopped me from merging this is that `sendForm()` was introduced to make it simple to send forms, regardless of the implementation used. By implementing this change, it would be necessary for plugin devs to change styles again, creating confusion.

For example, in the old days, `FormAPI` had a `sendToPlayer()` method. `sendForm()` eliminated the need for this, so now we had `player->sendForm(form)`. This change would a reversion back to those days, which calls into question why we added this method at all. It still exposes the protocol internals to plugins (by way of the JSON schema for forms, which might be different from version to version), so it's not really providing a lot of value other than just a shiny wrapper around `sendDataPacket(ModalFormRequestPacket::create(...))`.

The short of this is that I think removing this interface ruins the usability of this method, while also not being very useful for anything else.

**Comment by SOF3** [comment]
The usability of this method would actually be enhanced if we convert the callback parameter to some sort of Promise instead.

**Comment by dktapps** [comment]
I'm not sure how practical it would be to return a Promise for this without a builtin form implementation (we wouldn't be able to statically infer types otherwise).

**Comment by dktapps** [comment]
I'm not sure how practical it would be to return a Promise for this without a builtin form implementation (we wouldn't be able to statically infer types otherwise).

**Comment by SOF3** [comment]
Perhaps we keep the form interface and have it process the data then?

```php
/**
 * @phpstan-template R processed client data
 */
interface Form {
    public function getData();
    /**
     * @phpstan-return R
     */
    publc function processResponse($rawClientResponse);
}

class Player {
    /**
     * @phpstan-template R
     * @phpstan-param Form<R>
     * @phpstan-return Promise<R>
     */
    public function sendForm(Form $form) : Promise { ... }
}
```

**Comment by SOF3** [comment]
Is it true that this issue does not cater for client response validation? In that case, it might be meaningful to keep the interface after all.

**Comment by dktapps** [comment]
The response validation is done by the implementation.

**Comment by dktapps** [comment]
This PR is now heavily stale.

Considering the relatively tiny benefits offered by this PR, and the large scale disruption it would cause to existing plugins, I don't think it makes much sense to proceed.
